documentation/RLC-customization-doc.md

```diff
@@ -0,0 +1,167 @@
+# Customization on the RLC rest-level client libraries
+
+## Generate RLC Client
+
+Follow [quickstart](https://aka.ms/azsdk/rlc/js) to generate the rest-level client from OpenAPI specs.
+
+It's advised to put the generated code into the folder `generated`, add your customization code under the folder `src` and then export or re-export them as needed.
+
+```yaml
+source-code-folder-path: ./src/generated
+```
+
+## Custom authentication
+
+Some services require a custom authentication flow. For example Metrics Advisor uses Key Authentication, however MA requires 2 headers for key authentication `Ocp-Apim-Subscription-Key` and `x-api-key`, which is different to the usual key authentication which only requires a single key.
+
+In this case we customize as follows:
+
+1. Hand author a `PipelinePolicy` that takes values for both keys and sign the request
+2. Hand author a wrapping client factory function
+3. In the wrapping factory, we create a new client with the generated factory
+4. Inject the new policy to the client
+5. Return the client
+6. Only expose the wrapping factory and hide the generated factory.
+
+Here is an example in Metrics Advisor for the [wrapping function](https://github.com/MaryGao/azure-sdk-for-js/blob/metrix-advisor-poc/sdk/metricsadvisor/ai-metrics-advisor-rest/src/credentialHelper.ts#L10-L21) and [customized policy](https://github.com/MaryGao/azure-sdk-for-js/blob/metrix-advisor-poc/sdk/metricsadvisor/ai-metrics-advisor-rest/src/metricsAdvisorKeyCredentialPolicy.ts).
+
+With this user experience is the same as it is with any other RLC, as they just need to create a new client from the default exported factory function.
+
+```typescript
+import MetricsAdvisor, { paginate } from "@azure-rest/ai-metricsadvisor";
+
+const client = MetricsAdvisor("https://<endopoint>", {
+  key: "<apiKey>",
+  subscriptionKey: "<subscriptionKey>",
+});
+```
+
+## Custom paging helper
+
+Eventhough the code generator provides a pagination helper for RLCs, there are services that implement their own pagination pattern, different to the standard specification of `x-ms-pageable`.
+
+One example is the Metrics Advisor service, which implements a pagination pattern in which getting the next page can be called with `GET` or `POST` depending on the resource.
+
+The standard pagination pattern, assumes `GET` for getting the next pages. In this case we implemented a custom paginate helper that has the same public interface as the generated helper but under the hoods has an additional pagination implementation to use `POST`. Also this custom helper has an internal map that indicates which operations need `POST` and which need `GET`.
+
+You can find the [example code](https://github.com/MaryGao/azure-sdk-for-js/blob/metrix-advisor-poc/sdk/metricsadvisor/ai-metrics-advisor-rest/src/paginateHelper.ts#L25) to support the `POST` not `GET` method. The generated paging helper is hidden and the custom paginate helper is exposed.
```

Comment:
Please put the code here directly or find a place in azure-sdk-for-js repo, this is for public customer to read, thay may not have the access to your private forked repo.

Reply:
Sure